Don't share state between parts

OliviaYtterbrink opened the pull request.

OliviaYtterbrink committed:
Don't share state between parts

MishaDemianenko commented:
Where can i find tests for proposed changeset?

OliviaYtterbrink committed:
Handle thread safety in part creation

OliviaYtterbrink committed:
Stress test lazy cache

OliviaYtterbrink committed:
checkstyle

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexPopulator.java:
```diff
@@ -78,14 +80,16 @@ public synchronized void drop() throws IOException
+        for ( Map.Entry<CoordinateReferenceSystem,List<IndexEntryUpdate<?>>> entry : batchMap.entrySet() )
         {
-            part.applyUpdateBatch();
+            select( entry.getKey() ).add( entry.getValue() );
```
please correct me if i wrong here: on each batch of updates now we will try to take a lock to select correct populator from cache while that is required only first time when we need populator for that key.
As far as i can see particular populator accessed from one thread only (population job itself) - would it be possible to keep local cache of already selected populators, since those never change as soon as they created?

craigtaverner replied:
Something like a first loop over the batchMap.keySet() to call select on each CRS, and build a local crs->partPopulator map, and then do the loop over the entrySet and use the local map for the pre-selected partPopulators? I assume this is a performance optimization to minimize the total number of synchronizations done.

MishaDemianenko replied:
correct

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCache.java:
```diff
@@ -124,22 +142,22 @@ T select( ValueGroup valueGroup ) throws E
         {
         case DATE:
-            return date != null ? function.apply( date ) : orElse;
+            return parts[date.ordinal()] != null ? function.apply( parts[date.ordinal()] ) : orElse;
```
sync in this class is wrong you have readers that will not gonna see parts that are created; Instead of doing this kind of error-prone synchronization would recommend available sync collections. For example: CHM

sherfert replied:
What is CHM?

MishaDemianenko replied:
CHM -> ConcurrentHashMap

sherfert replied:
The reason we didn't go for that is the we wanted to ensure that `factory.newLocalDateTime()` (and similar) is only called once. And you should only pass side-effect-free things to `ConcurrentHashMap.computeIfAbsent` thus it can't guarantee that its only called once, if we don't add synchronization anyways. Is there an error in this logic?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/TemporalIndexPopulator.java:
```diff
@@ -73,13 +77,16 @@ public synchronized void drop() throws IOException
+        for ( Map.Entry<ValueGroup,List<IndexEntryUpdate<?>>> entry : batchMap.entrySet() )
         {
-            part.applyUpdateBatch();
+            select( entry.getKey() ).add( entry.getValue() );
```
same comment about select and locking inside as above

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCacheTest.java:
```diff
@@ -0,0 +1,144 @@
+
+        for ( CacheStresser stresser : stressers )
+        {
+            stresser.interrupt();
```
please use custom flag to stop the work instead of interrupts

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCacheTest.java:
```diff
@@ -0,0 +1,144 @@
+    {
+        SpatialIndexCache<String,Exception> cache;
+        Random r = new Random();
+        Exception failed;
```
instead of managing exceptions yourself would recommend to submit tasks to some pull and then join on futures that you will after task submissions

sherfert replied:
This is test only. Is that really that important?

MishaDemianenko replied:
well test code has the same importance as any other and i do not see a reason why that should be worse

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCacheTest.java:
```diff
@@ -0,0 +1,144 @@
+public class SpatialIndexCacheTest
+{
+    @Test
+    public void stressCache() throws Exception
```
in any concurrency releated tests its a good practice to set some timeout to make sure this will not gonna stuck ever

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/TemporalIndexCacheTest.java:
```diff
@@ -48,45 +62,153 @@ public void shouldIterateOverCreatedParts() throws Exception
     }
 
+    @Test
+    public void stressCache() throws Exception
```
same comments here as to other twin brother test

sherfert committed:
Make TemporalIndexCache use a ConcurrentHashMap

sherfert committed:
Remove generic exception from SpatialIndexCache

Use ExecutorService in stress tests

sherfert committed:
Make SpatialIndexCache use ConcurrentHashMap

For the time being we use CHM instead, since its better tested
compared to CopyOnWriteHashMap.

sherfert committed:
Merge remote-tracking branch 'origin/3.4' into 3.4-threadsafe-population-add

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCacheTest.java:
```diff
@@ -0,0 +1,149 @@
+public class SpatialIndexCacheTest
+{
+
+    @SuppressWarnings( "Duplicates" )
```
that's new level
